internal/typeinfo/info.go

```diff
@@ -9,110 +9,182 @@ import (
+				}
+				return nil, fmt.Errorf("two types found with name %q: %q and %q", t.Name(), dupeArg.typ().String(), t.String())
+			}
+			argInfo[t.Name()] = info
+		case reflect.Pointer:
+			return nil, fmt.Errorf("need struct or map, got pointer to %s", t.Elem().Kind())
+		default:
+			return nil, fmt.Errorf("need struct or map, got %s", t.Kind())
+		}
+	}
+	return argInfo, nil
 }
 
-func (si *structInfo) Typ() reflect.Type {
-	return si.structType
+// InputMember returns an input locator for a member of a struct or map.
+func (argInfo ArgInfo) InputMember(typeName string, memberName string) (Input, error) {
+	vl, err := argInfo.getMember(typeName, memberName)
+	if err != nil {
+		return nil, err
+	}
+	input, ok := vl.(Input)
+	if !ok {
+		return nil, fmt.Errorf("internal error: %s cannot be used as input", vl.ArgType().Kind())
+	}
+	return input, nil
 }
 
-func (si *structInfo) TypeMember(member string) (Member, error) {
-	tm, ok := si.tagToField[member]
+// OutputMember returns an output locator for a member of a struct or map.
+func (argInfo ArgInfo) OutputMember(typeName string, memberName string) (Output, error) {
+	vl, err := argInfo.getMember(typeName, memberName)
+	if err != nil {
+		return nil, err
+	}
+	output, ok := vl.(Output)
 	if !ok {
-		return nil, fmt.Errorf(`type %q has no %q db tag`, si.structType.Name(), member)
+		return nil, fmt.Errorf("internal error: %s cannot be used as output", vl.ArgType().Kind())
 	}
-	return tm, nil
+	return output, nil
 }
 
-func (si *structInfo) GetAllMembers() ([]Member, error) {
+// AllStructOutputs returns a list of output locators that locate every member
+// of the named type along with the names of the members. If the type is not a
+// struct an error is returned.
+func (argInfo ArgInfo) AllStructOutputs(typeName string) ([]Output, []string, error) {
+	arg, err := argInfo.getArg(typeName)
+	if err != nil {
+		return nil, nil, err
+	}
+	si, ok := arg.(*structInfo)
+	if !ok {
+		return nil, nil, fmt.Errorf("cannot generate columns for non-struct type")
+	}
 	if len(si.tags) == 0 {
-		return nil, fmt.Errorf(`no "db" tags found in struct %q`, si.structType.Name())
+		return nil, nil, fmt.Errorf(`no "db" tags found in struct %q`, si.structType.Name())
 	}
 
-	var tms []Member
+	var outputs []Output
 	for _, tag := range si.tags {
-		tms = append(tms, si.tagToField[tag])
+		outputs = append(outputs, si.tagToField[tag])
 	}
-	return tms, nil
+	return outputs, si.tags, nil
 }
 
-type mapInfo struct {
-	mapType reflect.Type
+// getArg finds information about a named arg type in argInfo.
+func (argInfo ArgInfo) getArg(typeName string) (arg, error) {
+	arg, ok := argInfo[typeName]
+	if !ok {
+		argNames := []string{}
+		for argName := range argInfo {
+			argNames = append(argNames, argName)
+		}
+		// Sort for consistant error messages.
+		sort.Strings(argNames)
+		return nil, typeMissingError(typeName, argNames)
+	}
+	return arg, nil
 }
```

Comment:
[Nit] What is your opinion on refactoring the error to a separate function (I know we already have `typeMissingError`), to change the code in each function to:
```go
        arg, ok := argInfo[typeName]
	if !ok {
		newFunctionForError(typeName, mapToList(argInfo))
	}
```
instead of:
```go
	arg, err := argInfo.getArg(typeName)
	if err != nil {
		return nil, nil, err
	}
```

Reply:
I think it could be nice, except that we have to do this code in the other place we call this function with a tiny difference, and we would probably also want to create a function there that's very similar but slightly different (the map keys are strings here but they are `reflect.Types` in the other callsite so we need to do `.Name()` there). Annoyingly generics can't really help here (we're not calling `String`, but `Name` which makes it awkward)

I've put a comment in both places this code appears to reduce the mental overhead of reading it but I think its fine how it is.